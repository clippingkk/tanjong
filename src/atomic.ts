import { ProfileQuery } from "./schema/generated"
import { atomWithMMKV } from "./utils/jotai"

export const tokenAtom = atomWithMMKV<string | null>('me:token', null)
export const uidAtom = atomWithMMKV<number | null>('me:uid', null)
export const profileAtom = atomWithMMKV<ProfileQuery['me'] | null>('me:profile', null)