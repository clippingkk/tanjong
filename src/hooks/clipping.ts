import { useMemo } from "react"
import { Dimensions } from "react-native"
import { Clipping } from "../schema/generated"

export function useClippingCellAvgHeight(cs: Pick<Clipping, 'content'>[]) {
  const itemSizeCellHeight = useMemo(() => {
    const recents = cs
    if (!recents) {
      return 10
    }
    const screenWidth = Dimensions.get('screen').width
    const availableWidth = screenWidth - 16 * 2 - 8 * 2
    const textsInOneLine = availableWidth / 14
    const lines = recents.reduce((acc, cur) => {
      const ls = cur.content.length / textsInOneLine
      acc += ls
      return acc
    }, 0)
    const avgLines = lines / recents.length
    // 行数 * line-height * font-size + paddingTop + paddingBottom
    const cellHeight = avgLines * 1.2 * 14 + 16 * 2
    return cellHeight
  }, [cs])
  return itemSizeCellHeight
}
