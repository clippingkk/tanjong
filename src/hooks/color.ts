import { useColorModeValue } from "native-base";

export function useTextColor(): string {
  return useColorModeValue('grey.100', 'grey.900')
}