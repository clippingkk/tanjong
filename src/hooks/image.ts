import { useState, useEffect } from "react"
import ImageColors from "react-native-image-colors"

export function useImagePrimaryColor(img: string) {
  const [bookImagePrimaryColor, setBookImagePrimaryColor] = useState<string | null>(null)
  useEffect(() => {
    if (!img) {
      return
    }

    ImageColors.getColors(img).then(res => {
      if (res.platform === 'android') {
        setBookImagePrimaryColor(res.vibrant!)
        return
      }
      if (res.platform === 'ios') {
        setBookImagePrimaryColor(res.primary!)
      }
    })
  }, [img])
  return bookImagePrimaryColor
}