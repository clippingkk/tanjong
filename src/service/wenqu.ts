import { Cache, Key, State } from "swr"
import { WENQU_SIMPLE_TOKEN, WENQU_API_HOST } from "../constants/config"

type WenquErrorResponse = {
  code: number
  error: string
}

export async function wenquRequest<T extends object>(url: string, options: RequestInit = {}): Promise<T> {
  options.headers = {
    ...(options.headers || {}),
    'X-Simple-Check': WENQU_SIMPLE_TOKEN
  }
  options.credentials = 'include'
  options.mode = 'cors'

  try {
    const response: T | WenquErrorResponse = await fetch(WENQU_API_HOST + url, options).then(res => res.json())
    if ('error' in response) {
      throw new Error(response.error)
    }
    return response
  } catch (e) {
    // Sentry.captureException(e)
    return Promise.reject(e)
  }
}

export interface WenquBook {
  id: number
  rating: number
  author: string
  pubdate: string
  totalPages: number
  originTitle: string
  image: string
  doubanId: number
  title: string
  url: string
  press: string
  isbn: string
  tags: string[]
  authorIntro: string
  summary: string
  createdAt: string
  updatedAt: string
}

export interface WenquSearchResponse {
  count: number
  books: WenquBook[]
}

export class WenquSWRCache implements Cache<WenquBook> {
  private _cache = new Map()

  keys(): IterableIterator<string> {
    return this._cache.keys()
  }
  get(key: Key): State<WenquBook, any> | undefined {
    throw new Error("Method not implemented.")
  }
  set(key: Key, value: State<WenquBook, any>): void {
    throw new Error("Method not implemented.")
  }
  delete(key: Key): void {
    throw new Error("Method not implemented.")
  }

}