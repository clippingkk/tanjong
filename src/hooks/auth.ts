import { useNavigation } from "@react-navigation/native"
import { useSetAtom } from "jotai"
import { useCallback } from "react"
import { tokenAtom, uidAtom } from "../atomic"
import { updateLocalToken } from "../utils/apollo"
// import SharedGroupPreferences from 'react-native-shared-group-preferences'
import { SharedGroupPreferencesKey } from "../constants/config"
import { setItem as WidgetKitSetItem, reloadAllTimelines as WidgetKitReloadAllTimelines } from 'react-native-widgetkit'
import { Platform } from "react-native"
import { useApolloClient } from "@apollo/client"
import { Toast } from "native-base"
const widgetAppIDKey = "app:my:id"
const widgetAppTokenKey = "app:token"
const widgetAppWidgetType = "app:widgetType"

export function usePostAuth() {
  const setToken = useSetAtom(tokenAtom)
  const setUid = useSetAtom(uidAtom)
  const nav = useNavigation()
  const client = useApolloClient()

  return useCallback(async (token: string, uid: number) => {
    setToken(token)
    updateLocalToken(token)
    setUid(uid)
    if (Platform.OS === 'ios') {
      // await SharedGroupPreferences.setItem(widgetAppIDKey, uid, SharedGroupPreferencesKey)
      // await SharedGroupPreferences.setItem(widgetAppTokenKey, token, SharedGroupPreferencesKey)
      WidgetKitSetItem(widgetAppIDKey, uid.toString(), SharedGroupPreferencesKey)
      WidgetKitSetItem(widgetAppTokenKey, token, SharedGroupPreferencesKey)
      WidgetKitSetItem(widgetAppWidgetType, "public", SharedGroupPreferencesKey)
      WidgetKitReloadAllTimelines()
    }

    Toast.show({
      title: 'Logged in'
    })
    await client.resetStore()
    if (nav.canGoBack()) {
      nav.goBack()
    }
  }, [setToken, setUid, nav])
}
