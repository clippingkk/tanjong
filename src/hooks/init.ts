import { useAtom, useSetAtom } from "jotai"
import { useEffect } from "react"
import { updateLocalToken } from "../utils/apollo"
import { profileAtom, tokenAtom } from "../atomic"
import { useProfileQuery } from "../schema/generated"
// import SplashScreen from 'react-native-splash-screen'

export function useOnInit() {
  const setProfile = useSetAtom(profileAtom)

  const np = useProfileQuery({
    variables: {
      id: 1,
    }
  })

  useEffect(() => {
    // 同步已登录用户数据
    if (!np.data?.me) {
      return
    }
    setProfile(np.data.me)
  }, [np.data?.me])

  useEffect(() => {
    // SplashScreen.hide()
  }, [])
}