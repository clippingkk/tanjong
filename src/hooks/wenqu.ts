import { useEffect, useMemo, useRef, useState } from "react"
import useSWR from 'swr'
import { WenquBook, wenquRequest, WenquSearchResponse } from "../service/wenqu"

export function useSingleBook(doubanId?: string | null, skip?: boolean) {
  return useSWR<WenquSearchResponse>(doubanId && `/books/search?dbId=${doubanId}`)
}

export function useBookSearch(query: string, offset: number) {
  const [books, setBooks] = useState<WenquSearchResponse | null>(null)
  useEffect(() => {
    if (query.length < 3) {
      return
    }

    wenquRequest<WenquSearchResponse>(`/books/search?query=${query}&limit=50&offset=${offset}`)
      .then(bs => {
        setBooks(bs)
      })
  }, [query, offset])

  return books
}