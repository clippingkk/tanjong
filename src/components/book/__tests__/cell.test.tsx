import 'react-native'
import React from 'react'
import { act, render, renderHook, screen, waitFor } from '@testing-library/react-native'
import { NativeBaseProvider } from 'native-base'
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'jotai'
import { uidAtom } from '../../../atomic'
import BookCell from '../cell'
import nock from 'nock'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { useSingleBook } from '../../../hooks/wenqu'
import { GluestackUIProvider } from '@gluestack-ui/themed'
import { config } from '@gluestack-ui/config'

test.only('book cell will rendered correctly', async () => {
  const qc = new QueryClient({
    defaultOptions: {
      queries: {
        staleTime: Infinity,
        retry: false,
      }
    }
  })
  const expectation = nock('https://wenqu.annatarhe.cn')
    .get('/api/v1/books/search?dbId=1111')
    .reply(200, {
      books: [{ image: 'hello', title: 'wwww' }],
      count: 2
    })

  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={qc}>
      <NativeBaseProvider initialWindowMetrics={{
        frame: { x: 0, y: 0, width: 0, height: 0 },
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
      }}>
        <Provider initialValues={[[uidAtom, 1]]}>
          <GluestackUIProvider config={config}>
            <NavigationContainer>
              {children}
              <BookCell bookDoubanID='1111' />
            </NavigationContainer>
          </GluestackUIProvider>
        </Provider>
      </NativeBaseProvider>
    </QueryClientProvider >
  )
  const { result } = renderHook(() => useSingleBook('1111'), {
    wrapper
  })
  await waitFor(() => {
    return expect(result.current.isSuccess).toBe(true)
  })
  // make sure every thing rendered
  await act(() => { })

  expect(result.current.data).toBeDefined()
  expect(screen).toMatchSnapshot()
  expectation.done()
})
