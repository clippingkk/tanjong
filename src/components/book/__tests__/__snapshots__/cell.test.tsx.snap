// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book cell will rendered correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "elevation": 7,
        "height": 250,
        "justifyContent": "center",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.29,
        "shadowRadius": 4.65,
      }
    }
  >
    <Text
      accessibilityRole="link"
      href="/BookDetail?book=%5Bobject%20Object%5D"
      onPress={[Function]}
    >
      <View
        dataSet={{}}
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "transparent",
              },
              {
                "aspectRatio": 1.7777777777777777,
                "height": 200,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "bottom": 0,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            >
              <ActivityIndicator />
            </View>
          </View>
        </View>
        <Text
          dataSet={{}}
          numberOfLines={2}
          style={
            {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": "LXGWWenKai-Regular",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 21,
              "marginTop": 8,
              "textAlign": "center",
              "textDecorationLine": undefined,
            }
          }
        >
          wwww
        </Text>
      </View>
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
